### PR TITLE
Add the ability to support custom config files

### PR DESCRIPTION
**Pull request by lucydodo**
This patch changes to support custom configuration file. This PR is related to issue #2317

Two options are added for the command line:
- `-c [config_file]`
- `--config [config_file]`

We can also use environment variables without using options on the command line.
`export DB4S_CONFIG_FILE=[config_file]`

If there is an environment variable and use the `-c` or `--config` option, the environment variable is ignored.

Usage
------
```
$ ./DB\ Browser\ for\ SQLite -c db4sconfig
$ ./DB\ Browser\ for\ SQLite --config db4sconfig
```

```
$ export DB4S_CONFIG_FILE=db4sconfig
$ ./DB\ Browser\ for\ SQLite
```

Possible problem
-----------------
~~The database file may be **overwritten** if the user enters only the value of the database file to be open without specifying a configuration file parameter when using this option. -> `$ ./DB\ Browser\ for\ SQLite --config veryImportantDB.db`~~
Solved through this commit(60ddbdd6bd7b9f4ed098da6cd8614bcc0fa2e8c9).

Known Bugs
------------
If you specify an environment variable on macOS and run it with the following command, the config file is not saved anywhere.
`open ./DB\ Browser\ for\ SQLite.app/`
However, if you specify as follows, it is saved normally.
`./DB\ Browser\ for\ SQLite.app/Contents/MacOS/DB\ Browser\ for\ SQLite`

Test Environment
-----------------
- macOS Catalina (10.15.6, 19G2021)

### CODE REVIEW

**Comment by lucydodo** [comment]
@mgrojo I have permission to merge, but I haven't finished studying and analyzing the project source tree yet.
May I ask for a review to you due to I am worried that it will cause unexpected problems to project via my code.

File: src/Application.cpp
Comment (mgrojo) [inline src/Application.cpp:28]: This would be more elegant with a setter and only the configuration file is needed. If it's empty, no configuration file has been set.

File: src/Application.cpp
Comment (mgrojo) [inline src/Application.cpp:37]: Should exit here? Note that in line 160 you assume an argument has been passed and skip it.

File: src/Application.cpp
Comment (mgrojo) [inline src/Application.cpp:40]: Consider giving the variable name so user knows what variable to unset to stop the warning.

File: src/Settings.cpp
Comment (mgrojo) [inline src/Settings.cpp:31]: In modern C++ is better to use `nullptr` instead of `NULL`.

File: src/Settings.cpp
Comment (mgrojo) [inline src/Settings.cpp:34]: To prevent the known bug, maybe `status()` might help. If the constructor doesn't write the file (which I hope it doesn't), checking status might be useful to avoid writing a file that is not in INI format. Maybe it should abort in that case, so it would be a good ide to check it in the setter.

File: src/Settings.cpp
Comment (mgrojo) [inline src/Settings.cpp:39]: Now that we use a pointer, it would be a good idea to construct the settings object only once. I think saving preferences would be much more efficient in that case, but I don't know why it worked like that before, maybe I'm missing something.

File: src/Settings.cpp
Comment (mgrojo) [inline src/Settings.cpp:63]: Why not using the usual `->` syntax?

File: src/Settings.h
Comment (mgrojo) [inline src/Settings.h:14]: If the setter is added, this would be of course in the private part. By the way, `...Dir` here seems like a leftover of a previous idea, it should be `...File`

File: src/Settings.cpp
Comment (mgrojo) [reply src/Settings.cpp:34]: A new file should be allowed, so it might be necessary to check also the existence.

**Review by mgrojo — CHANGES_REQUESTED** [review]
A good start! Thanks for letting me review it.

File: src/Settings.cpp
Comment (lucydodo) [reply src/Settings.cpp:34]: I tried to use `status()`, but it didn't help because QSettings constructor write the file first.
How about check first line starts with `[%General]` if the file already exist?

Comment (mgrojo) [reply src/Settings.cpp:34]: Strange design to overwrite the file in the constructor... Maybe using QSettings and setting later the filename and format?

If that won't work, your idea about checking the first line seems sensible.

Comment (lucydodo) [reply src/Settings.cpp:34]: As far as I know that is impossible because as soon as use QSettings the file is created or overwritten.

**Comment by lucydodo** [comment]
@mgrojo Thanks for the review! 😄  I modified the code according to your review.
In addition, I designed to ignore the option if user does NOT have permission to the file.

I added a comment, but it seems that the code is like 'spaghetti code'. I'll try think about neat code continue. 👀

File: src/Settings.cpp
Comment (mgrojo) [inline src/Settings.cpp:14]: This makes more sense as a static member variable as userConfigurationFile.

File: src/Application.cpp
Comment (mgrojo) [inline src/Application.cpp:163]: I'd unify these two messages in a simpler one:
qWarning() << qPrintable(tr("The -c/--config option requires an argument. The option is ignored."));

When it's ignored, it should be obvious that the application will run. Less work for translators :smiley:

File: src/Settings.cpp
Comment (mgrojo) [inline src/Settings.cpp:28]: For a clearer public interface, and for keeping it similar to the old one, I think this method should be made private. Then, it has to be called from every method using `settings`, like before, when we constructed a new object inside each method. All the code in the method should prevent double initialization checking that `settings` is equal to `nullptr` before initializing it.

File: src/Settings.cpp
Comment (mgrojo) [inline src/Settings.cpp:46]: If you're concerned about spaghetti code, you could take advantage of short-circuit evaluation of `&&` and unify all these conditions in one `if` (separated in different lines for readability).

File: src/Settings.cpp
Comment (mgrojo) [reply src/Settings.cpp:28]: Another option would be to make a single initialize method that can optionally pass a user configuration file and it has to be called before any other use of the static class. If the user userConfigurationFileArg parameter is QString() (default) the standard format is used. The issue with this solution is that if some call to any other method is added before this initialization, the application will crash (it's not a major problem, since it will be easily spotted). This can already happen, by the way. The issue with the other option is that, if a call to a setValue or a getValue is added in the future, before the call to setUserConfigurationFile, that value will be written or read from the standard location instead of the specified configuration file. It might be better the crash, because it will be spotted, but I don't have a strong opinion about it.

**Review by mgrojo — COMMENTED** [review]
From a user point of view would be ready for merging, but I've made some comments regarding the code which you might want to consider.

**Comment by lucydodo** [comment]
@mgrojo Thanks for the second review! 😄  I modified the code according to your review.

File: src/Settings.cpp
Comment (mgrojo) [inline src/Settings.cpp:41]: Just for better performance, if you add a check in `setSettingsObject()` covering all the code as follows:

```
if(!settings) {
...
}
```
then we would avoid reserving the memory, constructing the object, parsing the file or the registry, and finally deleting the object. It should work perfectly, although it had to be checked. I think it's ready for merging and the optimization can be done later, but I say it if you want to try.

**Review by mgrojo — APPROVED** [review]

File: src/Settings.cpp
Comment (lucydodo) [reply src/Settings.cpp:41]: Nice idea! I don't want anyone to be inconvenienced by my code. I'll fix it as soon as I get home. 😄

Comment (lucydodo) [reply src/Settings.cpp:45]: I just fixed the code. (4e45bce33db86c9776bb9db8335fefbcf1c1b902)

**Comment by mgrojo** [comment]
Thanks, @lucydodo. It's now merged!

After merging, I've seen something that I think could be improved. If the file doesn't exist, it should be allowed and passed to `QSettings` so an initial file can be populated. Another option could be having an Export Settings option from the GUI, but although we had that, I think it is better to allow to create a new file from the argument. Application will run with default values and the file will be initialized with whatever setting the user changes, and they will be able to reuse it in the future. What do you think?

**Comment by lucydodo** [comment]
@mgrojo Sorry I didn't understand what you say because my English is not very well.
You mean that the user give the uncreated file as an argument?

**Comment by mgrojo** [comment]
Sorry, I was executing another branch :smile: It works as expected, forget my comment.

**Pull request merged**
